lib/IRGen/MetadataRequest.cpp

```diff
@@ -2677,9 +2677,16 @@ emitMetadataAccessByMangledName(IRGenFunction &IGF, CanType type,
   unsigned mangledStringSize;
   std::tie(mangledString, mangledStringSize) =
     IGM.getTypeRef(type, CanGenericSignature(), MangledTypeRefRole::Metadata);
-  
-  assert(mangledStringSize < 0x80000000u
-         && "2GB of mangled name ought to be enough for anyone");
+
+  // Android AArch64 reserves the top byte of the address for memory tagging
+  // since Android 11, so only use the bottom 23 bits to store this size
+  // and the 24th bit to signal that there is a size.
+  if (IGM.Triple.isAndroid() && IGM.Triple.getArch() == llvm::Triple::aarch64)
+    assert(mangledStringSize < 0x00800001u &&
+           "8MB of mangled name ought to be enough for Android AArch64");
```

Comment:
Is the memory tag always nonzero? Could we have a scheme where the relative address to the mangled string has `00` in the top byte, and any tag in the top byte indicates the metadata pointer has been resolved?

Reply:
@buttaface?

Reply:
> Is the memory tag always nonzero? 

I don't think we can assume it is anything, it is simply reserved for the OS now.

> Could we have a scheme where the relative address to the mangled string has 00 in the top byte, and any tag in the top byte indicates the metadata pointer has been resolved?

I don't think it would be good to assume anything about the top byte, even though you may be right about what tagging scheme they end up using (currently it's just a static tag, but [they plan to use ARM's hardware MTE and some new scheme they come up with](https://github.com/android/ndk/issues/1653#issuecomment-1011378779)).

I only continue to check the top bit because it was easier to `AND` the new Android check for the eighth bit below than check one `OR` the other, and the result is the same.

I think this is better than trying to unify the scheme for all platforms, _unless_ you plan on turning on MTE for Apple platforms too, in which case this will eventually be needed there too.